packages/pg-v5/lib/psql.js

```diff
@@ -1,107 +1,247 @@
+  debug('after create tunnel')
+
+  const psql = execPSQL(options)
+  const cleanupSignalTraps = trapAndForwardSignalsToChildProcess(psql)
+
+  try {
+    debug('waiting for psql or tunnel to exit')
+    // wait for either psql or tunnel to exit;
+    // the important bit is that we ensure both processes are
+    // always cleaned up in the `finally` block below
+    await Promise.race([
+      waitForPSQLExit(psql),
+      tunnel.waitForClose()
+    ])
+  } catch (err) {
+    debug('wait for psql or tunnel error', err)
+    throw err
+  } finally {
+    debug('begin tunnel cleanup')
+    cleanupSignalTraps()
+    tunnel.close()
+    kill(psql, 'SIGKILL')
+    debug('end tunnel cleanup')
+  }
+}
+
+// a small wrapper around tunnel-ssh
+// so that other code doesn't have to worry about
+// whether there is or is not a tunnel
+class Tunnel {
+  constructor (bastionTunnel) {
+    this.bastionTunnel = bastionTunnel
+    this.events = new EventEmitter()
+  }
+
+  async waitForClose () {
+    if (this.bastionTunnel) {
+      try {
+        debug('wait for tunnel close')
+        await once(this.bastionTunnel, 'close')
+        debug('tunnel closed')
+      } catch (err) {
+        debug('tunnel close error', err)
+        throw new Error('Secure tunnel to your database failed')
       }
+    } else {
+      debug('no bastion required; waiting for fake close event')
+      await once(this.events, 'close')
     }
-    psqlArgs = psqlArgs.concat(['--set', 'sslmode=require'])
+  }
 
-    let psql = spawn('psql', psqlArgs, { env: dbEnv, stdio: 'inherit' })
-    handlePsqlError(reject, psql)
-    psql.on('close', (data) => {
-      resolve()
-    })
-  })
+  close () {
+    if (this.bastionTunnel) {
+      debug('close tunnel')
+      this.bastionTunnel.close()
+    } else {
+      debug('no tunnel necessary; sending fake close event')
+      this.events.emit('close', 0)
+    }
+  }
+
+  static async connect (db, tunnelConfig) {
+    const tunnel = await bastion.sshTunnel(db, tunnelConfig)
+    return new Tunnel(tunnel)
+  }
 }
 
-function * exec (db, query) {
-  let configs = bastion.getConfigs(db)
+async function exec (db, query) {
+  const configs = bastion.getConfigs(db)
+  const options = psqlQueryOptions(query, configs.dbEnv)
 
-  yield bastion.sshTunnel(db, configs.dbTunnelConfig)
-  return yield execPsql(query, configs.dbEnv)
+  return runWithTunnel(db, configs.dbTunnelConfig, options)
 }
 
 async function execFile (db, file) {
-  let configs = bastion.getConfigs(db)
+  const configs = bastion.getConfigs(db)
+  const options = psqlFileOptions(file, configs.dbEnv)
 
-  await bastion.sshTunnel(db, configs.dbTunnelConfig)
-  return execPsqlWithFile(file, configs.dbEnv)
+  return runWithTunnel(db, configs.dbTunnelConfig, options)
 }
 
-function * interactive (db) {
-  let name = db.attachment.name
-  let prompt = `${db.attachment.app.name}::${name}%R%# `
-  let configs = bastion.getConfigs(db)
+async function interactive (db) {
```

Comment:
these were changed to `async` functions to be consistent and to reduce our usage of `co` since the node versions we support (LTS 10, 12, 14 at time of this writing) support `async/await` syntax natively.